File: python/medium/Unique_Paths_II.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        step_list = []
        for i in range(len(obstacleGrid)):
            step_list.append([])
            for j in range(len(obstacleGrid[0])):
                if obstacleGrid[i][j] == 1:
                    step_list[i].append(0)
                elif i == 0:
                    if j != 0 and step_list[i][j-1] == 0:
                        step_list[i].append(0)
                    else:
                        step_list[i].append(1)
                elif j == 0:
                    if i != 0 and step_list[i-1][j] == 0:
                        step_list[i].append(0)
                    else:
                        step_list[i].append(1)
                else:
                    step_list[i].append(step_list[i-1][j]+step_list[i][j-1])
        return step_list[-1][-1]

File: cpp/medium/Unique_Paths_II.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int m = obstacleGrid.size();
        int n = obstacleGrid[0].size();
        if (obstacleGrid[0][0] == 1) obstacleGrid[0][0] = 0;
        else obstacleGrid[0][0] = 1;
        
        for(int j = 1; j < m; ++j) {
            if (obstacleGrid[j][0] == 1 || obstacleGrid[j-1][0] == 0) obstacleGrid[j][0] = 0;
            else obstacleGrid[j][0] = 1;
        }
        for(int i = 1; i < n; ++i) {
            if (obstacleGrid[0][i] == 1 || obstacleGrid[0][i-1] == 0) obstacleGrid[0][i] = 0;
            else obstacleGrid[0][i] = 1;
        }

        for(int j = 1; j < m; ++j) {
            for(int i = 1; i < n; ++i) {
                if (obstacleGrid[j][i] == 1) obstacleGrid[j][i] = 0;
                else obstacleGrid[j][i] = obstacleGrid[j-1][i] + obstacleGrid[j][i-1];
            }
        }
        return obstacleGrid[m-1][n-1];
    }
};
